readthedocs/storage/rclone.py

```diff
@@ -0,0 +1,185 @@
+"""
+Wrapper around the rclone command.
+
+See https://rclone.org/docs.
+"""
+
+import os
+import subprocess
+
+import structlog
+from django.utils._os import safe_join as safe_join_fs
+from storages.utils import safe_join
+
+log = structlog.get_logger(__name__)
+
+
+class BaseRClone:
+
+    """
+    RClone base class.
+
+    This class allows you to interact with an rclone remote without
+    a configuration file, the remote declaration and its options
+    are passed in the command itself.
+
+    This base class allows you to use the local file system as remote.
+
+    :param remote_type: You can see the full list of supported providers at
+     https://rclone.org/#providers.
+    :param rclone_bin: Binary name or path to the rclone binary.
+     Defaults to ``rclone``.
+    :param default_options: Options passed to the rclone command.
+    :parm env_vars: Environment variables used when executing the rclone command.
+     Useful to pass secrets to the command, since all arguments and options will be logged.
+    """
+
+    remote_type = None
+    rclone_bin = "rclone"
+    default_options = [
+        # Number of file transfers to run in parallel.
+        "--transfers=8",
+        # Skip based on checksum (if available) & size, not mod-time & size.
+        "--checksum",
+        "--verbose",
+    ]
+    env_vars = {}
+
+    # pylint: disable=no-self-use
+    def _build_target_path(self, path):
+        """Build the final target path for the remote."""
+        return path
+
+    def build_target(self, path):
+        """
+        Build the proper target using the current remote type.
+
+        We start the remote with `:` to create it on the fly,
+        instead of having to create a configuration file.
+        See https://rclone.org/docs/#backend-path-to-dir.
+
+        :param path: Path to the remote target.
+        """
+        path = self._build_target_path(path)
+        return f":{self.remote_type}:{path}"
+
+    def execute(self, action, args, options=None):
+        """
+        Execute an rclone subcommand.
+
+        :param action: Name of the subcommand.
+        :param list args: List of positional arguments passed the to command.
+        :param list options: List of options passed to the command.
+        """
+        options = options or []
+        command = [
+            self.rclone_bin,
+            action,
+            *self.default_options,
+            *options,
+            "--",
+            *args,
+        ]
+        env = os.environ.copy()
+        env.update(self.env_vars)
+        log.info("Executing rclone command.", command=command)
+        log.debug("env", env=env)
+        result = subprocess.run(
+            command,
+            capture_output=True,
+            env=env,
+            # TODO: Fail or let the caller decide what to do?
+            check=True,
```

Comment:
Not sure about this one, we could manually check the exit code or catch the exception

Reply:
I'm not sure to understand the doubt here. What are the different scenarios and their pros/cons?

Reply:
If the command exists with a status different than 0 with `check=True` it will raise an exception, if false (the default), then it won't raise an exception and continue normally (but you can still see check for the exit code manually)